server/services/v1/search_indexer.go

```diff
@@ -276,27 +337,75 @@ func UnpackSearchFields(doc map[string]interface{}, collection *schema.DefaultCo
 	return searchKey, tableData, doc, nil
 }
 
-func FlattenObjects(data map[string]any) map[string]any {
+// FlattenObjectsAndArray flattens objects and array of objects. Primitive arrays or an array of arrays remain untouched. The format of flattening is as follows:
+//
+//	Object,
+//	  - {"a": {"b": "c}} will be flattened as {"a.b": "c}
+//	Array of Object,
+//	  - {"arr_obj": [{"a": 1, "b": "first"}, {"a": 2, "b": "second"}]" will be flattened as {"arr_obj.a":[1, 2], "arr_obj.b":["first","second"]}
+//	Primitive arrays inside an object or array of objects follow the same rule as above,
+//	  - {"arr_obj": [{"a": [1, 2, 3]}, {"a": [4, 5]}]} will be flattened as {"arr_obj.a": [[1, 2, 3], [4, 5]]}
+func FlattenObjectsAndArray(data map[string]any, mergeWithArray func(name string) bool) map[string]any {
 	resp := make(map[string]any)
-	flattenObjects("", data, resp)
+	flattenObjectsAndArray("", data, resp, false, mergeWithArray)
 	return resp
 }
 
-func flattenObjects(key string, obj map[string]any, resp map[string]any) {
+func flattenObjectsAndArray(key string, obj map[string]any, resp map[string]any, isArray bool, mergeWithArray func(name string) bool) {
 	if key != "" {
 		key += schema.ObjFlattenDelimiter
 	}
 
+	assign := func(result map[string]any, k string, v any, isArray bool) {
```

Comment:
We can simplify this to:
```
      // get or convert existing value to array                                    
      var arr []any                                                        
      if e, ok := result[k]; ok {                                           
          if arr, ok = e.([]any); !ok {                             
              arr = []any{e}                                        
          }                                                                        
      }                                                                            
                                                                                   
      if vArr, ok := v.([]any); ok && mergeWithArray(k) {                          
          // if it is an array, and it needs to be flattened as array, otherwise store as-is
          result[k] = append(arr, vArr...)                                 
      } else {                                                                     
          result[k] = append(arr, v)                                       
      }                                                           
```